Prevent integer width conversion in caml_int_compare

smuenzel-js opened the pull request:
Previously, in 64bit mode, res was computed as 32bits, and then extended to 64bits.
This patch eliminates the sign-extension instruction.

smuenzel-js committed:
Prevent integer width conversion in caml_int_compare

Previously, in 64bit mode, res was computed as 32bits, and then extended to 64bits.
This patch eliminates the sign-extension instruction.

gasche commented:
Could you include a comment in the code to record the intent in this change?

xclerc commented:
This code pattern has a number of occurrences,
*e.g.* `caml_int32_compare_unboxed`.
Shouldn't we update all of them? (Possibly by
defining a macro.)

xavierleroy commented:
Does this change improves performance in a measurable way?  int32 -> int64 conversions are extremely cheap, and the proposed code is less readable than the current code...  I'm not asking for big speedups, just a hint that it makes a non-negligible difference.

smuenzel-js committed:
Update to use macro for compare

smuenzel-js commented:
I've updated the code to use a macro, and changed all the other comparisons that had this issue.

Benchmark results on a loop that adds up the results of comparisons:
```
New: 
./a.out  15.00s user 0.00s system 99% cpu 15.011 total
./a.out  14.86s user 0.01s system 99% cpu 14.874 total
./a.out  14.94s user 0.00s system 99% cpu 14.947 total

Old:
./a.out  17.76s user 0.01s system 99% cpu 17.767 total
./a.out  17.30s user 0.00s system 99% cpu 17.316 total
./a.out  18.09s user 0.01s system 99% cpu 18.100 total
```

This is the program:
```
let do_compare (x : int) (y : int) =
  let x = Sys.opaque_identity x in
  let y = Sys.opaque_identity y in
  compare x y

let () =
  let result = ref 0 in
  for i = 0 to 4_000_000_000 do
    result := !result + (do_compare 432 5);
    result := !result + (do_compare 5 432);
    result := !result + (do_compare 432 432);
  done;
  print_int !result
```

The difference in the two functions is clear when you look at the disassembly.
Previously:
```
XDIS 0000000000418630: BASE 31C0                     xor eax, eax
XDIS 0000000000418632: BASE 4839F7                   cmp rdi, rsi
XDIS 0000000000418635: BASE 0F9FC0                   setnle al
XDIS 0000000000418638: BASE 31D2                     xor edx, edx
XDIS 000000000041863a: BASE 4839F7                   cmp rdi, rsi
XDIS 000000000041863d: BASE 0F9CC2                   setl dl
XDIS 0000000000418640: BASE 29D0                     sub eax, edx
XDIS 0000000000418642: LONGMODE 4898                     cdqe 
XDIS 0000000000418644: BASE 488D440001               lea rax, ptr [rax+rax*1+0x1]
XDIS 0000000000418649: BASE C3                       ret 
```
With this patch:
```
XDIS 0000000000418430: BASE 31C0                     xor eax, eax
XDIS 0000000000418432: BASE 4839F7                   cmp rdi, rsi
XDIS 0000000000418435: BASE 0F9CC2                   setl dl
XDIS 0000000000418438: BASE 0F9FC0                   setnle al
XDIS 000000000041843b: BASE 0FB6D2                   movzx edx, dl
XDIS 000000000041843e: BASE 4829D0                   sub rax, rdx
XDIS 0000000000418441: BASE 488D440001               lea rax, ptr [rax+rax*1+0x1]
XDIS 0000000000418446: BASE C3                       ret 
```

smuenzel-js committed:
Add changelog entry

xclerc commented:
Thanks for the changes; Is there a reason to not apply it
to `{int32,int64,nativeint}_cmp`?

xclerc commented:
Answering my own question: the aforementioned functions
do not need to be modified because they return an `int`.

xavierleroy commented on runtime/floats.c:
```diff
@@ -1008,8 +1008,15 @@ intnat caml_float_compare_unboxed(double f, double g)
+  intnat equal_g = g == g;
+  intnat res = greater - less + equal_f - equal_g;
+  return res;
 }
```
Do you really need to write it one asm instruction per line?  What's wrong with
```
(intnat)(f > g) - (intnat)(f < g) + (intnat)(f == f) - (intnat)(g == g)
```

xavierleroy commented on runtime/ints.c:
```diff
@@ -25,6 +25,14 @@
+  intnat __compare_greater = __compare_v1 > __compare_v2; \
+  intnat __compare_less    = __compare_v1 < __compare_v2; \
+  intnat result = __compare_greater - __compare_less
+
```
I'm not a fan of macros that declare variables in the current scope.  (As opposed to within a block.)  I'm not sure macros are needed anyway, see below.

smuenzel-js replied:
A macro is nice because there's one canonical reference for the code, not a couple of places where it's inlined. That helps with maintenance -- only one place to change, and it's also discoverable by those new to the codebase.
And we also don't have to document the reasoning for the code in more than one place.

I'd be happy to change the style of the macro, but I'm not exactly sure how.
One way could be make the entire function the macro, and have something like
`DECLARE_COMPARE_UNBOXED(caml_int32_compare_unboxed,int32_t)`
Another would be to take the one-liner you propose below, and make that into a macro.

xclerc replied:
The suggestion, if I understand it correctly, was to do something akin to:

```
#define COMPARE_INT(type, v1, v2, result) \
  { type __compare_v1 = v1; \
    type __compare_v2 = v2; \
    intnat __compare_greater = __compare_v1 > __compare_v2; \
    intnat __compare_less    = __compare_v1 < __compare_v2; \
    intnat result = __compare_greater - __compare_less }
```
following the style of *e.g.* https://github.com/ocaml/ocaml/blob/trunk/runtime/bigarray.c#L195.

I think it is a bit safer/better than your change in https://github.com/ocaml/ocaml/pull/2250/commits/1df410b987fdb3b5dedbc2c3083a13b656b1c144, because
we are sure the expressions passed to the macro are not evaluated
mutilple times.

xavierleroy commented on runtime/ints.c:
```diff
@@ -126,7 +134,7 @@ CAMLprim value caml_bswap16(value v)
 {
-  int res = (v1 > v2) - (v1 < v2);
+  COMPARE_INT(value, v1, v2, res);
   return Val_int(res);
```
`res` is of type `intnat`, so it should be `Val_long(res)`.

xavierleroy commented on runtime/ints.c:
```diff
@@ -126,7 +134,7 @@ CAMLprim value caml_bswap16(value v)
 CAMLprim value caml_int_compare(value v1, value v2)
 {
-  int res = (v1 > v2) - (v1 < v2);
+  COMPARE_INT(value, v1, v2, res);
```
Macro-less alternative:
```
   intnat res = (intnat)(v1 > v2) - (intnat)(v1 < v2);
```

xavierleroy commented on runtime/ints.c:
```diff
@@ -314,7 +322,8 @@ CAMLprim value caml_int32_to_float(value v)
 intnat caml_int32_compare_unboxed(int32_t i1, int32_t i2)
 {
-  return (i1 > i2) - (i1 < i2);
+  COMPARE_INT(int32_t, i1, i2, res);
```
Likewise.

xavierleroy commented on runtime/ints.c:
```diff
@@ -562,7 +571,8 @@ CAMLprim value caml_int64_to_nativeint(value v)
 intnat caml_int64_compare_unboxed(int64_t i1, int64_t i2)
 {
-  return (i1 > i2) - (i1 < i2);
+  COMPARE_INT(int64_t,i1,i2,res);
```
Likewise.

xavierleroy commented on runtime/ints.c:
```diff
@@ -824,7 +834,8 @@ CAMLprim value caml_nativeint_to_int32(value v)
 intnat caml_nativeint_compare_unboxed(intnat i1, intnat i2)
 {
-  return (i1 > i2) - (i1 < i2);
+  COMPARE_INT(intnat, i1, i2, res);
```
Likewise.

xavierleroy reviewed (commented):
Thanks for the benchmark numbers, which are convincing at least for x86.

Some suggestions below from someone who doesn't like macros.

smuenzel-js committed:
Simplify expressions.

xavierleroy commented:
I liked it better before the last commit.  Could we please stop suggesting things and pushing code, and instead proceed with a proper review?  Which means that the code should not change every day.

smuenzel-js commented:
@xavierleroy I've reverted the latest change.

xavierleroy committed:
Expand comment on COMPARE_INT macro

xavierleroy committed:
Make spacing uniform in invocations of COMPARE_INT macro

The house style is one space after comma.

xavierleroy reviewed (approved):
I pushed directly some small changes (documentation comment, spacing).  Looks OK to me.

xavierleroy committed:
Merge branch 'trunk' into patch-1
